sandbox/lua/decoders/json.lua

```diff
@@ -0,0 +1,187 @@
+
+- Timestamp (string, optional, default nil)
+    string specifying json field to map to message Timestamp,
+    expects field value to be unix epoch timestamp in seconds.
+
+.. code-block:: javascript
+
+    {
+      "msg": "Start Request",
+      "event": "artemis.web.ensure-running",
+      "extra": {
+        "workspace-id": "cN907xLngi"
+      },
+      "time": "2015-05-06T20:40:05.509926234Z",
+      "severity": 1
+    }
+
+*Example Heka Configuration*
+
+.. code-block:: ini
+
+    [ArtemisLogInput]
+    type = "LogstreamerInput"
+    log_directory = "/srv/artemis/current/logs"
+    file_match = 'artemis\.log'
+    decoder = "JsonDecoder"
+
+    [JsonDecoder]
+    type = "SandboxDecoder"
+    filename = "lua_decoders/json.lua"
+
+        [JsonDecoder.config]
+        type = "artemis"
+        payload_keep = true
+        map_fields = true
+        Severity = "severity"
+
+*Example Heka Message*
+
+:Timestamp: 2015-05-06 20:40:05 -0800 PST
+:Type: artemis
+:Hostname: test.example.com
+:Pid: 0
+:UUID: 8e414f01-9d7f-4a48-a5e1-ae92e5954df5
+:Payload:
+:EnvVersion:
+:Severity: 1
+:Fields:
+    | name:"msg" value_type:STRING value_string:"Start Request"
+    | name:"event" value_type:STRING value_string:"artemis.web.ensure-running"
+    | name:"extra.workspace-id" value_type:STRING value_string:"cN907xLngi"
+    | name:"time" value_type:STRING value_string:"2015-05-06T20:40:05.509926234Z"
+--]]
+
+require "cjson"
+local util = require("util")
+
+local payload_keep = read_config("payload_keep")
+local map_fields   = read_config("map_fields")
+
+local field_map = {
+    Payload    = read_config("Payload"),
+    Uuid       = read_config("Uuid"),
+    Type       = read_config("Type"),
+    Logger     = read_config("Logger"),
+    Hostname   = read_config("Hostname"),
+    Severity   = read_config("Severity"),
+    EnvVersion = read_config("EnvVersion"),
+    Pid        = read_config("Pid"),
+    Timestamp  = read_config("Timestamp")
+}
+
+local field_type_map = {
+    Payload    = "string",
+    Uuid       = "string",
+    Type       = "string",
+    Logger     = "string",
+    Hostname   = "string",
+    Severity   = "number",
+    EnvVersion = "number",
+    Pid        = "number",
+    Timestamp  = "number"
+}
+
+local msg = {
+    Payload    = nil,
+    Uuid       = nil,
+    Type       = read_config("type") or "json",
+    Logger     = nil,
+    Hostname   = nil,
+    Severity   = nil,
+    EnvVersion = nil,
+    Pid        = nil,
+    Timestamp  = nil,
+    Fields     = nil
+}
+
+function process_message()
+    local ok, json = pcall(cjson.decode, read_message("Payload"))
+    if not ok then
+        return -1, "Failed to decode JSON."
+    end
+
+    -- keep payload, or not
+    if payload_keep then
+        msg.Payload = read_message("Payload")
+    end
+
+    -- map fields
+    if map_fields then
+        for F, f in pairs(field_map) do
+            if type(json[f]) == field_type_map[F] then
+                if F == "Timestamp" then
```

Comment:
Yes, sorry the docs could use quite some improvement here. We provide a date_time module in the Lua sandbox that allows you to parse date strings if you provide the correct strftime format representation string. Calling [date_time.build_strftime_grammar](https://github.com/mozilla-services/lua_sandbox/blob/v0.7.0/modules/date_time.lua#L217) returns an LPEG grammar that does the parsing for you. It returns a Lua table that can be turned into ns-since-epoch using [date_time.time_to_ns](https://github.com/mozilla-services/lua_sandbox/blob/v0.7.0/modules/date_time.lua#L13). You can add an optional timestamp_format config setting where users can specify the strftime format string. Lack of a timestamp_format setting would presume the default timestamp format of ns-since-epoch, i.e. exactly what we have now.